tweet_recommendations/other_methods/dbscan_based_method.py

```diff
@@ -0,0 +1,143 @@
+
+        epsilon = np.mean(self.neighbours.query(embeddings, k=neighbours_count, n_jobs=mp.cpu_count())[0])
+
+        if self.verbose:
+            print(f"Found epsilon: {epsilon}")
+            print("Fitting DBSCAN ... ")
+
+        self.clusterizer = DBSCAN(metric='manhattan', eps=epsilon, min_samples=MIN_SAMPLES, n_jobs=mp.cpu_count())
+        x["cluster_label"] = self.clusterizer.fit_predict(embeddings)
+
+        if self.verbose:
+            print("Clustering finished.")
+
+        x["hashtags"] = x["hashtags"].apply(lambda r: [elem["text"] for elem in r])
+
+        grouped_clusters = x.groupby(["cluster_label"])
+
+        if self.verbose:
+            print(f"Samples grouped into {len(np.unique(self.clusterizer.labels_))} clusters")
+            print("Building clusters' data")
+
+        self._centroids_data = []
+        self._corresponding_to_centroids_data_hashtags = []
+
+        for cluster_number, cluster in tqdm.tqdm(grouped_clusters, total=len(np.unique(x["cluster_label"])),
+                                                 disable=not self.verbose):
+            cluster_embeddings = np.vstack(cluster["embedding"].to_numpy())
+            cluster_center = np.mean(cluster_embeddings, axis=0).reshape(1, -1)
+            distances = distance.cdist(cluster_center, cluster_embeddings)[0]  # matrix of shape (1 x N) is returned
+
+            cluster_medoid_index = np.argmin(
+                distances)  # we are sure that it will not point to itself because mean center should be not existent
+
+            cluster_medoid = cluster_embeddings[cluster_medoid_index]
+            aggregated_hashtags = list(
+                chain.from_iterable(cluster["hashtags"].to_numpy()))
+            sorted_most_popular_hashtags = [hashtag for hashtag, count in
+                                            Counter(aggregated_hashtags).most_common()]
+
+            self._centroids_data.append(cluster_medoid)
+            self._corresponding_to_centroids_data_hashtags.append(sorted_most_popular_hashtags)
+
+        self._centroids_data = np.vstack(self._centroids_data)
+        self._corresponding_to_centroids_data_hashtags = np.asarray(self._corresponding_to_centroids_data_hashtags)
+
+        return self
+
+    def transform(self, x: Union[List[List[str]], List[str]]) -> np.ndarray:
+        """
+        For a given tweet/tweets embeddings recommend hashtags.
+        :param x: list of list of str or list of str. If first argument of x is a list is str, it is assumed that list
+            contains already lemmatized text. If single str is present as first element, it is assumed
+            that lemmatization has to be performed.
+        :return: Iterable of recommended hashtags.
+        """
+        embeddings = self.embedding_method.transform(x)
+        embeddings = embeddings if len(embeddings.shape) == 2 else embeddings.reshape(1, -1)
+        distances = cosine_distances(embeddings, self._centroids_data)
```

Comment:
Wcześniej: manhattan,
potem: euclidean,
teraz: cosine
tak ma być?

Reply:
Manhattan:

> The DBSCANalgorithm has two parameters: epsilon (The radius of our neighborhoods around a data point p) and minPoints(the minimum number of data points we want in a neighborhood to define a cluster). In our case, we use the Manhattan distance as metric (See eq 2), and for each tweet

Euclidean: to samo, co wcześniej, czyli w sumie nie wiadomo:

>The fact of grouping vectors in this manner is recognized as vector quantization. To perform this, we detect the closest feature vector to the clusters center for each cluster, this vector is considered as the centroid of the cluster.

Cosine: 

>ccordingly, we calculate distance between the entered tweet vector and centroids of clusters obtained in the previous step using the cosine similarity, if the entered tweet vector representation is near a given centroid, then this tweet share semantic and syntactic contexts with the larger part of tweets in this cluster.

Reply:
Jak wcześniej: zamieniłbym Euklidesa na Manhattan.